In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
stark_clicks = pd.read_csv('stark_clicks.csv')
stark_clicks.head()

Raw User Id  \
0  6F556808-A569-4568-B4E1-DA47CFC77784   
1  0A28682E-C502-439F-80D7-8E08B1372D81   
2  EEFDF90E-6F2D-409B-BD25-D781632002C0   
3  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
4  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   

                                           Unique Id  Epoch Timestamp  
0  68546823-1544918255-682-1-23-29-f41xG-rsdlCMGX...       1544918344  
1  166592346-1544917712-438-1-40-30-SiMgB-200ef81...       1544918311  
2  447583387-1544918173-290-1-16-24-wKX4p-ec19228...       1544918396  
3  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
4  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074

In [3]:
stark_clicks.isnull().sum()

Raw User Id        0
Unique Id          0
Epoch Timestamp    0
dtype: int64

In [4]:
# Check for duplicate rows
duplicate_rows = stark_clicks.duplicated()

# Select the duplicate rows
duplicate_rows_stark_clicks = stark_clicks[duplicate_rows]

# Display the duplicate rows
duplicate_rows_stark_clicks

Raw User Id  \
4      3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
5      3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
6      3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
7      3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
8      3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
...                                     ...   
11114  142C8D26-87F3-4AAB-BD39-68E7ED986679   
11122  8D33ACBC-D6F1-4430-8EEB-04370C68C3FA   
11123  8D33ACBC-D6F1-4430-8EEB-04370C68C3FA   
11133  10D74A19-8345-4D4C-A517-10628B4DA093   
11134  10D74A19-8345-4D4C-A517-10628B4DA093   

                                               Unique Id  Epoch Timestamp  
4      901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
5      901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
6      901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
7      901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
8      901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
...                                                  ...              ...  
11114  433108580-1545001523-631-1-12-15-OH66T-hBKpDRS...       1545001600  
11122  797303747-1545001856-219-1-17-18-aFTaB-k4f4OeZ...       1545001980  
11123  797303747-1545001856-219-1-17-18-aFTaB-k4f4OeZ...       1545001980  
11133  168328554-1545002907-150-1-17-5-g2zaG-u9Ud1JvL...       1545003064  
11134  168328554-1545002907-150-1-17-5-g2zaG-u9Ud1JvL...       1545003064  

[3747 rows x 3 columns]

In [5]:
duplicate_rows_stark_clicks.head()

Raw User Id  \
4  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
5  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
6  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
7  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   
8  3787D06D-BBB6-4FB1-8EED-A9A53F7C74D3   

                                           Unique Id  Epoch Timestamp  
4  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
5  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
6  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
7  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074  
8  901417839-1544918008-930-1-18-5-LdZaR-gDbB690l...       1544918074

In [6]:
stark_clicks.duplicated().value_counts()

False    7400
True     3747
dtype: int64

In [7]:
stark_clicks = stark_clicks.drop_duplicates()

In [8]:
stark_clicks.shape

(7400, 3)

In [9]:
Clicks = 1
stark_clicks['Clicks'] = Clicks

In [10]:
grouped_stark_clicks = stark_clicks.groupby(['Unique Id','Epoch Timestamp','Raw User Id']).agg({'Clicks': 'sum'})
grouped_stark_clicks

Clicks
Unique Id                                          Epoch Timestamp Raw User Id                                 
100149385-1545078590-251-1-17-0-OJyIZ-G4qW82Al7... 1545079430      8EE141F1-7B16-4BC1-AF4E-664CECC4AB19       1
100259881-1544902281-609-1-13-4-9Ufui-7d5f8259-... 1544902282      E626EB4B-2EFA-46A2-83D9-4DD16B729C47       1
100295427-1544825932-074-1-15-16-iuqzg-28331f8d... 1544825932      83E1F189-8773-41B7-BFEE-33DD9570F464       1
100430032-1545048679-355-1-17-20-5Doaa-hVVfTkl7... 1545048813      C4CCC776-A9AF-4D86-8690-9DDBDBDDD560       1
101023807-1544820215-000-1-25-8-uGzFD-4GZYfmxxn... 1544820287      A9CC1907-4E8B-44E0-8CA8-A5F05B3BE56B       1
...                                                                                                         ...
999009114-1545072498-761-1-10-26-au6UA-b5ba3d52... 1545072555      E3109131-D3DA-494C-8DEA-0D519E9420B0       1
999159358-1544852322-191-1-14-28-NwZEz-94efd599... 1544852323      459ED855-AB12-4466-8932-C9394B243621       1
999370847-1544810505-087-1-11-5-6uXbQ-b2286ee0-... 1544810518      828B1A7D-85ED-4BC2-A190-CF84213533CE       1
999660891-1545085422-888-1-15-25-xbpw1-597afb97... 1545085528      89FE8C61-E480-417C-8906-A01BB2E27C7C       1
999847078-1545054639-552-1-24-35-dLFAg-J4lTh0LL... 1545054693      8E871120-8A9A-4F9D-AD49-7C05A280D0CC       1

[7400 rows x 1 columns]

In [11]:
grouped_stark_clicks.reset_index(inplace=True)
grouped_stark_clicks

Unique Id  Epoch Timestamp  \
0     100149385-1545078590-251-1-17-0-OJyIZ-G4qW82Al...       1545079430   
1     100259881-1544902281-609-1-13-4-9Ufui-7d5f8259...       1544902282   
2     100295427-1544825932-074-1-15-16-iuqzg-28331f8...       1544825932   
3     100430032-1545048679-355-1-17-20-5Doaa-hVVfTkl...       1545048813   
4     101023807-1544820215-000-1-25-8-uGzFD-4GZYfmxx...       1544820287   
...                                                 ...              ...   
7395  999009114-1545072498-761-1-10-26-au6UA-b5ba3d5...       1545072555   
7396  999159358-1544852322-191-1-14-28-NwZEz-94efd59...       1544852323   
7397  999370847-1544810505-087-1-11-5-6uXbQ-b2286ee0...       1544810518   
7398  999660891-1545085422-888-1-15-25-xbpw1-597afb9...       1545085528   
7399  999847078-1545054639-552-1-24-35-dLFAg-J4lTh0L...       1545054693   

                               Raw User Id  Clicks  
0     8EE141F1-7B16-4BC1-AF4E-664CECC4AB19       1  
1     E626EB4B-2EFA-46A2-83D9-4DD16B729C47       1  
2     83E1F189-8773-41B7-BFEE-33DD9570F464       1  
3     C4CCC776-A9AF-4D86-8690-9DDBDBDDD560       1  
4     A9CC1907-4E8B-44E0-8CA8-A5F05B3BE56B       1  
...                                    ...     ...  
7395  E3109131-D3DA-494C-8DEA-0D519E9420B0       1  
7396  459ED855-AB12-4466-8932-C9394B243621       1  
7397  828B1A7D-85ED-4BC2-A190-CF84213533CE       1  
7398  89FE8C61-E480-417C-8906-A01BB2E27C7C       1  
7399  8E871120-8A9A-4F9D-AD49-7C05A280D0CC       1  

[7400 rows x 4 columns]

In [12]:
grouped_stark_clicks.head()

Unique Id  Epoch Timestamp  \
0  100149385-1545078590-251-1-17-0-OJyIZ-G4qW82Al...       1545079430   
1  100259881-1544902281-609-1-13-4-9Ufui-7d5f8259...       1544902282   
2  100295427-1544825932-074-1-15-16-iuqzg-28331f8...       1544825932   
3  100430032-1545048679-355-1-17-20-5Doaa-hVVfTkl...       1545048813   
4  101023807-1544820215-000-1-25-8-uGzFD-4GZYfmxx...       1544820287   

                            Raw User Id  Clicks  
0  8EE141F1-7B16-4BC1-AF4E-664CECC4AB19       1  
1  E626EB4B-2EFA-46A2-83D9-4DD16B729C47       1  
2  83E1F189-8773-41B7-BFEE-33DD9570F464       1  
3  C4CCC776-A9AF-4D86-8690-9DDBDBDDD560       1  
4  A9CC1907-4E8B-44E0-8CA8-A5F05B3BE56B       1

In [13]:
RawUserId_Clicks=grouped_stark_clicks.groupby(['Raw User Id']).agg({'Clicks': 'sum'})

In [14]:
RawUserId_Clicks

Clicks
Raw User Id                                 
00076F3F-AC69-4CE6-8067-CFA84572D115       1
00121FD8-17D8-40BF-8EBF-DA7BD2107C39       1
002BA8AD-D8ED-464D-8FD2-3B980AAAB280       1
003B24B2-10C3-4325-9131-C9527E787595       1
0049FC92-0859-4DC1-AC29-8B68868304FA       1
...                                      ...
FFB52702-AEE3-4969-BAB1-C9842C35ADC8       1
FFD6B446-9BC9-4610-8D26-6E45B0102B94       3
FFD9FAEB-BBDB-4E58-94E5-B94ACB140299       1
FFDFC01B-840C-4D7C-B973-41612C9C67C0       1
FFF811AB-0E7B-4DEF-9E46-94DBBC7F963A       1

[4026 rows x 1 columns]

In [15]:
RawUserId_Clicks.reset_index(inplace=True)
RawUserId_Clicks

Raw User Id  Clicks
0     00076F3F-AC69-4CE6-8067-CFA84572D115       1
1     00121FD8-17D8-40BF-8EBF-DA7BD2107C39       1
2     002BA8AD-D8ED-464D-8FD2-3B980AAAB280       1
3     003B24B2-10C3-4325-9131-C9527E787595       1
4     0049FC92-0859-4DC1-AC29-8B68868304FA       1
...                                    ...     ...
4021  FFB52702-AEE3-4969-BAB1-C9842C35ADC8       1
4022  FFD6B446-9BC9-4610-8D26-6E45B0102B94       3
4023  FFD9FAEB-BBDB-4E58-94E5-B94ACB140299       1
4024  FFDFC01B-840C-4D7C-B973-41612C9C67C0       1
4025  FFF811AB-0E7B-4DEF-9E46-94DBBC7F963A       1

[4026 rows x 2 columns]

In [16]:
stark_impressions=pd.read_csv('stark_impressions.csv',dtype={'App Id':'str'})

In [17]:
stark_impressions.duplicated().value_counts()

False    278075
True       1846
dtype: int64

In [18]:
#Drop Duplicates
stark_impressions = stark_impressions.drop_duplicates()

In [19]:
stark_impressions.shape

(278075, 10)

In [20]:
stark_impressions.head()

REQ_COUNTRY_ID                                  Raw User Id      App Id  \
0             246  raw_id:74A0C6FC-F8C6-450E-87AF-188BB71337E0  1436008834   
1             246  raw_id:A11F8505-010B-4479-9D15-8A56010E272E  1436008834   
2             246  raw_id:496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
3             246  raw_id:496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
4              93  raw_id:CA0BA044-FB83-4288-AB8D-1B15BF66ABE8   962892510   

   EXCHANGE_ID  DEVICE_OS  CAMPAIGN_ID  CREATIVE_ID_1  \
0           23         12       134686       50250928   
1           23         12       134686       50250936   
2        10044         12       134687       50250940   
3        10044         12       134687       50250936   
4        10049         12       134687       50250929   

                                           Unique Id  Epoch Timestamp  \
0  561103834-1544763382-669-1-35-19-R9ZTz-77b2c69...       1544763675   
1  200773481-1544766511-928-1-38-24-iw5CK-fa0d78f...       1544766534   
2  687624111-1544764246-009-1-23-27-YQwcX-YxdeBrC...       1544764471   
3  147773246-1544764478-510-1-16-12-sIxgj-GZz3YQN...       1544764503   
4  80896666-1544764807-846-1-18-25-9vYBl-21b88a77...       1544764911   

   IAB_CATEGORY_ID1  
0              11.0  
1              11.0  
2               NaN  
3               NaN  
4             325.0

In [21]:
Impressions = 1
stark_impressions['impressions'] = Impressions

In [22]:
stark_impressions.head()

REQ_COUNTRY_ID                                  Raw User Id      App Id  \
0             246  raw_id:74A0C6FC-F8C6-450E-87AF-188BB71337E0  1436008834   
1             246  raw_id:A11F8505-010B-4479-9D15-8A56010E272E  1436008834   
2             246  raw_id:496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
3             246  raw_id:496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
4              93  raw_id:CA0BA044-FB83-4288-AB8D-1B15BF66ABE8   962892510   

   EXCHANGE_ID  DEVICE_OS  CAMPAIGN_ID  CREATIVE_ID_1  \
0           23         12       134686       50250928   
1           23         12       134686       50250936   
2        10044         12       134687       50250940   
3        10044         12       134687       50250936   
4        10049         12       134687       50250929   

                                           Unique Id  Epoch Timestamp  \
0  561103834-1544763382-669-1-35-19-R9ZTz-77b2c69...       1544763675   
1  200773481-1544766511-928-1-38-24-iw5CK-fa0d78f...       1544766534   
2  687624111-1544764246-009-1-23-27-YQwcX-YxdeBrC...       1544764471   
3  147773246-1544764478-510-1-16-12-sIxgj-GZz3YQN...       1544764503   
4  80896666-1544764807-846-1-18-25-9vYBl-21b88a77...       1544764911   

   IAB_CATEGORY_ID1  impressions  
0              11.0            1  
1              11.0            1  
2               NaN            1  
3               NaN            1  
4             325.0            1

In [23]:
# Replace the exact string 'raw_id:' with an empty string in the 'Raw User Id' column
stark_impressions['Raw User Id'] = stark_impressions['Raw User Id'].replace('^raw_id:', '', regex=True)

# Display the updated DataFrame
print(stark_impressions)

        REQ_COUNTRY_ID                           Raw User Id      App Id  \
0                  246  74A0C6FC-F8C6-450E-87AF-188BB71337E0  1436008834   
1                  246  A11F8505-010B-4479-9D15-8A56010E272E  1436008834   
2                  246  496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
3                  246  496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
4                   93  CA0BA044-FB83-4288-AB8D-1B15BF66ABE8   962892510   
...                ...                                   ...         ...   
279916             246  917D11BF-2C20-45E9-BAB8-0E8DAE3EC484  1047961826   
279917             246  F1D1118C-C716-43B8-BE99-33BC5BD82D7B  1047961826   
279918             246  48049270-2B59-4023-BE7A-909BEA87A7B9  1047961826   
279919             246  0D77F6A2-4290-4B09-B30F-77D36ED9BD6B  1047961826   
279920             246  0C81060C-C869-4ACD-BA82-D469E2941F38  1047961826   

        EXCHANGE_ID  DEVICE_OS  CAMPAIGN_ID  CREATIVE_ID_1  \
0                23      

In [24]:
stark_impressions.head()

REQ_COUNTRY_ID                           Raw User Id      App Id  \
0             246  74A0C6FC-F8C6-450E-87AF-188BB71337E0  1436008834   
1             246  A11F8505-010B-4479-9D15-8A56010E272E  1436008834   
2             246  496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
3             246  496F124B-96AB-485B-9B2E-2C8910FFD510  1303389751   
4              93  CA0BA044-FB83-4288-AB8D-1B15BF66ABE8   962892510   

   EXCHANGE_ID  DEVICE_OS  CAMPAIGN_ID  CREATIVE_ID_1  \
0           23         12       134686       50250928   
1           23         12       134686       50250936   
2        10044         12       134687       50250940   
3        10044         12       134687       50250936   
4        10049         12       134687       50250929   

                                           Unique Id  Epoch Timestamp  \
0  561103834-1544763382-669-1-35-19-R9ZTz-77b2c69...       1544763675   
1  200773481-1544766511-928-1-38-24-iw5CK-fa0d78f...       1544766534   
2  687624111-1544764246-009-1-23-27-YQwcX-YxdeBrC...       1544764471   
3  147773246-1544764478-510-1-16-12-sIxgj-GZz3YQN...       1544764503   
4  80896666-1544764807-846-1-18-25-9vYBl-21b88a77...       1544764911   

   IAB_CATEGORY_ID1  impressions  
0              11.0            1  
1              11.0            1  
2               NaN            1  
3               NaN            1  
4             325.0            1

In [25]:
final_stark_impressions = stark_impressions.groupby(['Unique Id','Epoch Timestamp','Raw User Id']).agg({'impressions': 'sum'})
final_stark_impressions

impressions
Unique Id                                          Epoch Timestamp Raw User Id                                      
100001716-1544999799-668-1-17-28-o5c3X-WwejDHkn... 1544999869      8F40571F-758F-499B-9168-77A16CE85AA2            1
100003653-1544836022-429-1-13-28-hijMU-a3c4b65e... 1544841799      EE37511F-77C7-4DA8-BDFF-987586A182C8            1
100003711-1544806642-517-1-15-17-5B5sk-4a0e8070... 1544806700      FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5            1
100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7G... 1544784348      BD9A5622-3B1B-48F9-845E-5F0ACBC64620            1
100018996-1544922000-537-1-15-31-wXBmr-8b72497e... 1544922007      9B429BD4-3440-415A-BBF7-D7FA59FE1FDD            1
...                                                                                                              ...
999990176-1544879358-992-1-26-4-ekn17-0da5535a-... 1544879787      50392414-0A38-46AE-A4D2-FEFCB29E76BB            1
999990803-1544936355-070-1-28-36-9RJR9-4sf6b0yn... 1544936563      449B99AD-6014-4187-AD37-1AB45D0E0BE3            1
999994234-1544816621-387-1-16-14-ddAd0-Q42t2dcR... 1544816684      A8563E35-7386-4704-A749-1909004A921E            1
999997023-1544758994-380-1-10-30-BnR5X-7cbf6152... 1544759340      9162A1BE-5A52-4BC8-9A59-E754CC4BF491            1
999998726-1544916354-628-1-23-0-mejDA-jdXDf928T... 1544916443      39D9FB5A-090A-4CBB-9380-B20FBBEAC5E2            1

[278061 rows x 1 columns]

In [26]:
final_stark_impressions['impressions'].sum()

278075

In [27]:
final_stark_impressions.reset_index(inplace=True)
final_stark_impressions

Unique Id  Epoch Timestamp  \
0       100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1       100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2       100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3       100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4       100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   
...                                                   ...              ...   
278056  999990176-1544879358-992-1-26-4-ekn17-0da5535a...       1544879787   
278057  999990803-1544936355-070-1-28-36-9RJR9-4sf6b0y...       1544936563   
278058  999994234-1544816621-387-1-16-14-ddAd0-Q42t2dc...       1544816684   
278059  999997023-1544758994-380-1-10-30-BnR5X-7cbf615...       1544759340   
278060  999998726-1544916354-628-1-23-0-mejDA-jdXDf928...       1544916443   

                                 Raw User Id  impressions  
0       8F40571F-758F-499B-9168-77A16CE85AA2            1  
1       EE37511F-77C7-4DA8-BDFF-987586A182C8            1  
2       FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5            1  
3       BD9A5622-3B1B-48F9-845E-5F0ACBC64620            1  
4       9B429BD4-3440-415A-BBF7-D7FA59FE1FDD            1  
...                                      ...          ...  
278056  50392414-0A38-46AE-A4D2-FEFCB29E76BB            1  
278057  449B99AD-6014-4187-AD37-1AB45D0E0BE3            1  
278058  A8563E35-7386-4704-A749-1909004A921E            1  
278059  9162A1BE-5A52-4BC8-9A59-E754CC4BF491            1  
278060  39D9FB5A-090A-4CBB-9380-B20FBBEAC5E2            1  

[278061 rows x 4 columns]

In [28]:
stark_installs=pd.read_csv('stark_installs.csv')

In [29]:
stark_installs.head()

Unique Id  is_valid_unique_id  can_claim  \
0                               NaN                   0          0   
1                           BSylpKk                   0          0   
2                               NaN                   0          0   
3                               NaN                   0          0   
4  44e563b5ab59da90dfa0ba05b09c8e12                   0          0   

  country_code                           Raw User Id DEVICE_OS  \
0           US  babe71dc-c7a9-4689-bb77-f025a21851f1   android   
1           GT  d2dad60c-2542-4e02-bcb5-36fd1be20524   android   
2           PT  227519e8-6e50-4921-95e9-5d76731bcbed   android   
3           ID  98648ade-81a5-4e99-b183-40654cfd05ce   android   
4           US  53F94CA3-60BA-4EA7-A70E-0C503E11ED48       ios   

   Epoch Timestamp  
0       1544831396  
1       1544831407  
2       1544831397  
3       1544831403  
4       1544831404

In [30]:
filtered_stark_installs = stark_installs[stark_installs['can_claim'] != 0]

In [31]:
filtered_stark_installs.head()

Unique Id  is_valid_unique_id  \
493   257578208-1544831549-197-1-28-19-Uyglf-tkhDeh4...                   1   
597   681396195-1544820933-559-1-14-1-h7PWt-3a8de62e...                   1   
644   129915915-1544831848-220-1-37-31-ucycl-b6e8e79...                   1   
925   290058047-1544766550-167-1-24-27-HQzD3-SRO6gyp...                   1   
1598  975101916-1544830212-422-1-27-30-kJy4Q-ph15DDq...                   1   

      can_claim country_code                           Raw User Id DEVICE_OS  \
493           1           US  3628C029-DA16-4E50-B2A6-D82E8E97467B       ios   
597           1           US  4AEC6FE7-23F0-439A-9727-5DB28AEEDBAF       ios   
644           1           US  F77C685F-A32D-4DCD-9D36-343CFC8F381C       ios   
925           1           US  2F724DDA-992C-401B-BADB-2ADBC34F36BD       ios   
1598          1           US  A3229765-24B4-4D9D-8E28-1FC4B064DDC9       ios   

      Epoch Timestamp  
493        1544831846  
597        1544831933  
644        1544831991  
925        1544830723  
1598       1544831362

In [32]:
installs = 1
filtered_stark_installs['installs'] = installs

C:\Users\smita.s\AppData\Local\Temp\ipykernel_23320\2568825921.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_stark_installs['installs'] = installs


In [33]:
final_stark_installs = filtered_stark_installs.groupby(['Unique Id','Epoch Timestamp','Raw User Id']).agg({'installs': 'sum'})
final_stark_installs

installs
Unique Id                                          Epoch Timestamp Raw User Id                                   
100149385-1545078590-251-1-17-0-OJyIZ-G4qW82Al7... 1545081348      8EE141F1-7B16-4BC1-AF4E-664CECC4AB19         1
106311245-1544811114-298-1-17-16-8SJoH-XvDaBm32... 1544811425      4C1536FF-95A1-41D3-AD56-15626670ACAD         1
10631328-1544811668-756-1-10-24-Lr04P-e0766a83-... 1544815225      069B4B54-E624-4CEA-879A-3E010F436BBB         1
106827692-1544930730-111-1-10-0-d5dzT-67b07fe2-... 1544941226      E1E86C0B-35A2-4C9F-833D-33557A4AB653         1
106898838-1545055563-243-1-18-4-3vLxB-7S5OjAj4I... 1545055907      BFBBADD2-7068-459B-9AC1-FAE20127958E         1
...                                                                                                           ...
997149665-1544804392-994-1-37-18-QyiO0-7e795c07... 1544804846      974EBD50-A505-4303-8BF7-CB9B0519A5AD         1
998004542-1544845095-358-1-18-18-MMrPm-DLIZwgOZ... 1544846544      8497132E-35D0-41A5-B7E4-3816CE4DBD4C         1
998135497-1545078820-478-1-26-14-BqyXG-ZhPTKygP... 1545143588      94B23708-C511-478D-AF4F-C74619133FB4         1
999660891-1545085422-888-1-15-25-xbpw1-597afb97... 1545085684      89FE8C61-E480-417C-8906-A01BB2E27C7C         1
999911856-1544816414-845-1-23-8-XhWYg-wuSyciroo... 1544816786      EBFB26BF-9D58-41BB-8EEE-735E7E4B5FBF         1

[1200 rows x 1 columns]

In [34]:
final_stark_installs['installs'].sum()

1200

In [35]:
final_stark_installs.reset_index(inplace=True)
final_stark_installs

Unique Id  Epoch Timestamp  \
0     100149385-1545078590-251-1-17-0-OJyIZ-G4qW82Al...       1545081348   
1     106311245-1544811114-298-1-17-16-8SJoH-XvDaBm3...       1544811425   
2     10631328-1544811668-756-1-10-24-Lr04P-e0766a83...       1544815225   
3     106827692-1544930730-111-1-10-0-d5dzT-67b07fe2...       1544941226   
4     106898838-1545055563-243-1-18-4-3vLxB-7S5OjAj4...       1545055907   
...                                                 ...              ...   
1195  997149665-1544804392-994-1-37-18-QyiO0-7e795c0...       1544804846   
1196  998004542-1544845095-358-1-18-18-MMrPm-DLIZwgO...       1544846544   
1197  998135497-1545078820-478-1-26-14-BqyXG-ZhPTKyg...       1545143588   
1198  999660891-1545085422-888-1-15-25-xbpw1-597afb9...       1545085684   
1199  999911856-1544816414-845-1-23-8-XhWYg-wuSyciro...       1544816786   

                               Raw User Id  installs  
0     8EE141F1-7B16-4BC1-AF4E-664CECC4AB19         1  
1     4C1536FF-95A1-41D3-AD56-15626670ACAD         1  
2     069B4B54-E624-4CEA-879A-3E010F436BBB         1  
3     E1E86C0B-35A2-4C9F-833D-33557A4AB653         1  
4     BFBBADD2-7068-459B-9AC1-FAE20127958E         1  
...                                    ...       ...  
1195  974EBD50-A505-4303-8BF7-CB9B0519A5AD         1  
1196  8497132E-35D0-41A5-B7E4-3816CE4DBD4C         1  
1197  94B23708-C511-478D-AF4F-C74619133FB4         1  
1198  89FE8C61-E480-417C-8906-A01BB2E27C7C         1  
1199  EBFB26BF-9D58-41BB-8EEE-735E7E4B5FBF         1  

[1200 rows x 4 columns]

In [36]:
RawUserId_installs=final_stark_installs.groupby(['Raw User Id']).agg({'installs': 'sum'})

In [37]:
RawUserId_installs.reset_index(inplace=True)
RawUserId_installs

Raw User Id  installs
0     002BA8AD-D8ED-464D-8FD2-3B980AAAB280         1
1     003B24B2-10C3-4325-9131-C9527E787595         1
2     005CEE68-08B4-4C60-AA0B-838FB548AA3B         1
3     00777D3F-D3B4-4A63-B6D7-40CBDA43547B         1
4     00BE4A0C-C7C2-40C9-AC3A-C81CE4666104         1
...                                    ...       ...
1195  FF2D3490-B7F8-44EC-AB7B-ABE20D0C08E0         1
1196  FF35FBC9-881C-4BCA-9206-462FB71A2717         1
1197  FF895764-630F-4AA3-9530-5C35C037C834         1
1198  FFA791A2-82F8-40B7-AC53-3E9EE6F20DD1         1
1199  FFE0B5D6-F81B-4605-9B13-34E372810ED5         1

[1200 rows x 2 columns]

In [38]:
Impressions_Clicks = pd.merge(final_stark_impressions,grouped_stark_clicks, on=['Unique Id','Raw User Id','Epoch Timestamp'], how='outer')

In [39]:
Impressions_Clicks.head()

Unique Id  Epoch Timestamp  \
0  100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1  100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2  100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3  100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4  100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   

                            Raw User Id  impressions  Clicks  
0  8F40571F-758F-499B-9168-77A16CE85AA2          1.0     NaN  
1  EE37511F-77C7-4DA8-BDFF-987586A182C8          1.0     NaN  
2  FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5          1.0     NaN  
3  BD9A5622-3B1B-48F9-845E-5F0ACBC64620          1.0     NaN  
4  9B429BD4-3440-415A-BBF7-D7FA59FE1FDD          1.0     NaN

In [40]:
Impressions_Clicks_sum = Impressions_Clicks[['impressions', 'Clicks']].sum()

In [41]:
Impressions_Clicks_sum

impressions    278075.0
Clicks           7400.0
dtype: float64

In [42]:
All_Merged=pd.merge(Impressions_Clicks,final_stark_installs, on=['Unique Id','Epoch Timestamp','Raw User Id'], how='outer')

In [43]:
All_Merged.head()

Unique Id  Epoch Timestamp  \
0  100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1  100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2  100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3  100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4  100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   

                            Raw User Id  impressions  Clicks  installs  
0  8F40571F-758F-499B-9168-77A16CE85AA2          1.0     NaN       NaN  
1  EE37511F-77C7-4DA8-BDFF-987586A182C8          1.0     NaN       NaN  
2  FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5          1.0     NaN       NaN  
3  BD9A5622-3B1B-48F9-845E-5F0ACBC64620          1.0     NaN       NaN  
4  9B429BD4-3440-415A-BBF7-D7FA59FE1FDD          1.0     NaN       NaN

In [44]:
All_Merged_sum = All_Merged[['impressions', 'Clicks','installs']].sum()

In [45]:
All_Merged_sum

impressions    278075.0
Clicks           7400.0
installs         1200.0
dtype: float64

In [46]:
All_Merged_copy=All_Merged

In [47]:
All_Merged_copy

Unique Id  Epoch Timestamp  \
0       100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1       100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2       100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3       100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4       100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   
...                                                   ...              ...   
285915  997149665-1544804392-994-1-37-18-QyiO0-7e795c0...       1544804846   
285916  998004542-1544845095-358-1-18-18-MMrPm-DLIZwgO...       1544846544   
285917  998135497-1545078820-478-1-26-14-BqyXG-ZhPTKyg...       1545143588   
285918  999660891-1545085422-888-1-15-25-xbpw1-597afb9...       1545085684   
285919  999911856-1544816414-845-1-23-8-XhWYg-wuSyciro...       1544816786   

                                 Raw User Id  impressions  Clicks  installs  
0       8F40571F-758F-499B-9168-77A16CE85AA2          1.0     NaN       NaN  
1       EE37511F-77C7-4DA8-BDFF-987586A182C8          1.0     NaN       NaN  
2       FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5          1.0     NaN       NaN  
3       BD9A5622-3B1B-48F9-845E-5F0ACBC64620          1.0     NaN       NaN  
4       9B429BD4-3440-415A-BBF7-D7FA59FE1FDD          1.0     NaN       NaN  
...                                      ...          ...     ...       ...  
285915  974EBD50-A505-4303-8BF7-CB9B0519A5AD          NaN     NaN       1.0  
285916  8497132E-35D0-41A5-B7E4-3816CE4DBD4C          NaN     NaN       1.0  
285917  94B23708-C511-478D-AF4F-C74619133FB4          NaN     NaN       1.0  
285918  89FE8C61-E480-417C-8906-A01BB2E27C7C          NaN     NaN       1.0  
285919  EBFB26BF-9D58-41BB-8EEE-735E7E4B5FBF          NaN     NaN       1.0  

[285920 rows x 6 columns]

In [48]:
All_Merged_copy = All_Merged_copy.fillna({'Clicks': 0, 'impressions': 0, 'installs': 0})

In [49]:
All_Merged_copy

Unique Id  Epoch Timestamp  \
0       100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1       100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2       100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3       100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4       100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   
...                                                   ...              ...   
285915  997149665-1544804392-994-1-37-18-QyiO0-7e795c0...       1544804846   
285916  998004542-1544845095-358-1-18-18-MMrPm-DLIZwgO...       1544846544   
285917  998135497-1545078820-478-1-26-14-BqyXG-ZhPTKyg...       1545143588   
285918  999660891-1545085422-888-1-15-25-xbpw1-597afb9...       1545085684   
285919  999911856-1544816414-845-1-23-8-XhWYg-wuSyciro...       1544816786   

                                 Raw User Id  impressions  Clicks  installs  
0       8F40571F-758F-499B-9168-77A16CE85AA2          1.0     0.0       0.0  
1       EE37511F-77C7-4DA8-BDFF-987586A182C8          1.0     0.0       0.0  
2       FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5          1.0     0.0       0.0  
3       BD9A5622-3B1B-48F9-845E-5F0ACBC64620          1.0     0.0       0.0  
4       9B429BD4-3440-415A-BBF7-D7FA59FE1FDD          1.0     0.0       0.0  
...                                      ...          ...     ...       ...  
285915  974EBD50-A505-4303-8BF7-CB9B0519A5AD          0.0     0.0       1.0  
285916  8497132E-35D0-41A5-B7E4-3816CE4DBD4C          0.0     0.0       1.0  
285917  94B23708-C511-478D-AF4F-C74619133FB4          0.0     0.0       1.0  
285918  89FE8C61-E480-417C-8906-A01BB2E27C7C          0.0     0.0       1.0  
285919  EBFB26BF-9D58-41BB-8EEE-735E7E4B5FBF          0.0     0.0       1.0  

[285920 rows x 6 columns]

In [50]:
# Fill null values with a default value (e.g., 0)
All_Merged_copy['Epoch Timestamp'].fillna(0, inplace=True)

# Convert epoch timestamp to datetime
All_Merged_copy['Date_Time'] = pd.to_datetime(All_Merged_copy['Epoch Timestamp'], unit='s')

# Display the resulting DataFrame
All_Merged_copy

Unique Id  Epoch Timestamp  \
0       100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1       100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2       100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3       100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4       100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   
...                                                   ...              ...   
285915  997149665-1544804392-994-1-37-18-QyiO0-7e795c0...       1544804846   
285916  998004542-1544845095-358-1-18-18-MMrPm-DLIZwgO...       1544846544   
285917  998135497-1545078820-478-1-26-14-BqyXG-ZhPTKyg...       1545143588   
285918  999660891-1545085422-888-1-15-25-xbpw1-597afb9...       1545085684   
285919  999911856-1544816414-845-1-23-8-XhWYg-wuSyciro...       1544816786   

                                 Raw User Id  impressions  Clicks  installs  \
0       8F40571F-758F-499B-9168-77A16CE85AA2          1.0     0.0       0.0   
1       EE37511F-77C7-4DA8-BDFF-987586A182C8          1.0     0.0       0.0   
2       FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5          1.0     0.0       0.0   
3       BD9A5622-3B1B-48F9-845E-5F0ACBC64620          1.0     0.0       0.0   
4       9B429BD4-3440-415A-BBF7-D7FA59FE1FDD          1.0     0.0       0.0   
...                                      ...          ...     ...       ...   
285915  974EBD50-A505-4303-8BF7-CB9B0519A5AD          0.0     0.0       1.0   
285916  8497132E-35D0-41A5-B7E4-3816CE4DBD4C          0.0     0.0       1.0   
285917  94B23708-C511-478D-AF4F-C74619133FB4          0.0     0.0       1.0   
285918  89FE8C61-E480-417C-8906-A01BB2E27C7C          0.0     0.0       1.0   
285919  EBFB26BF-9D58-41BB-8EEE-735E7E4B5FBF          0.0     0.0       1.0   

                 Date_Time  
0      2018-12-16 22:37:49  
1      2018-12-15 02:43:19  
2      2018-12-14 16:58:20  
3      2018-12-14 10:45:48  
4      2018-12-16 01:00:07  
...                    ...  
285915 2018-12-14 16:27:26  
285916 2018-12-15 04:02:24  
285917 2018-12-18 14:33:08  
285918 2018-12-17 22:28:04  
285919 2018-12-14 19:46:26  

[285920 rows x 7 columns]

In [51]:
All_Merged_copy.head()

Unique Id  Epoch Timestamp  \
0  100001716-1544999799-668-1-17-28-o5c3X-WwejDHk...       1544999869   
1  100003653-1544836022-429-1-13-28-hijMU-a3c4b65...       1544841799   
2  100003711-1544806642-517-1-15-17-5B5sk-4a0e807...       1544806700   
3  100010818-1544784283-449-1-17-9-LGG5E-J17ksfX7...       1544784348   
4  100018996-1544922000-537-1-15-31-wXBmr-8b72497...       1544922007   

                            Raw User Id  impressions  Clicks  installs  \
0  8F40571F-758F-499B-9168-77A16CE85AA2          1.0     0.0       0.0   
1  EE37511F-77C7-4DA8-BDFF-987586A182C8          1.0     0.0       0.0   
2  FE6BB5EA-22E3-4F16-AAE5-5C3C324C3AA5          1.0     0.0       0.0   
3  BD9A5622-3B1B-48F9-845E-5F0ACBC64620          1.0     0.0       0.0   
4  9B429BD4-3440-415A-BBF7-D7FA59FE1FDD          1.0     0.0       0.0   

            Date_Time  
0 2018-12-16 22:37:49  
1 2018-12-15 02:43:19  
2 2018-12-14 16:58:20  
3 2018-12-14 10:45:48  
4 2018-12-16 01:00:07

In [52]:
# Export the DataFrame to a CSV file
All_Merged_copy.to_csv('All_Merged_copy.csv', index=False)

In [53]:
from IPython.display import FileLink

# Assuming you have a DataFrame called All_Merged_copy_Idwise_sum that you want to export
All_Merged_copy.to_csv('All_Merged_copy.csv', index=False)

# Create a download link for the CSV file
FileLink('All_Merged_copy.csv')

C:\Users\smita.s\All_Merged_copy.csv

In [54]:
All_Raw_User_IDs_Summarized_By_Installs=All_Merged_copy.groupby(['Raw User Id']).agg({'installs': 'sum'})
All_Raw_User_IDs_Summarized_By_Installs

installs
Raw User Id                                   
00000481-7D49-46DF-AF21-A8A2AAD60930       0.0
000020F8-F3AF-41F8-8896-82F5415E4ACB       0.0
0001826B-3F99-4612-81B3-DB7CDD60AEC8       0.0
0003FA5B-77BB-4F02-B625-0F16C990E304       0.0
00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       0.0
...                                        ...
f4b4af4e-6eb6-4dd0-a6da-cd7c1caa3dbc       0.0
f69dade6-5469-45e2-9799-df163940c0ab       0.0
f7baef29-3123-46d3-87ad-45a9f14a8227       0.0
fa5c0287-8958-45d0-997a-cd591b483756       0.0
fc85a4d7-e7fc-4794-8f49-f617fad8821f       0.0

[122073 rows x 1 columns]

In [55]:
All_Raw_User_IDs_Summarized_By_Installs.reset_index(inplace=True)
All_Raw_User_IDs_Summarized_By_Installs

Raw User Id  installs
0       00000481-7D49-46DF-AF21-A8A2AAD60930       0.0
1       000020F8-F3AF-41F8-8896-82F5415E4ACB       0.0
2       0001826B-3F99-4612-81B3-DB7CDD60AEC8       0.0
3       0003FA5B-77BB-4F02-B625-0F16C990E304       0.0
4       00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       0.0
...                                      ...       ...
122068  f4b4af4e-6eb6-4dd0-a6da-cd7c1caa3dbc       0.0
122069  f69dade6-5469-45e2-9799-df163940c0ab       0.0
122070  f7baef29-3123-46d3-87ad-45a9f14a8227       0.0
122071  fa5c0287-8958-45d0-997a-cd591b483756       0.0
122072  fc85a4d7-e7fc-4794-8f49-f617fad8821f       0.0

[122073 rows x 2 columns]

In [56]:
Install_OR_NoInstall=All_Raw_User_IDs_Summarized_By_Installs.groupby(['installs']).agg({'Raw User Id': 'count'})
Install_OR_NoInstall

Raw User Id
installs             
0.0            120873
1.0              1200

In [57]:
# Calculate the column total
column_total = Install_OR_NoInstall['Raw User Id'].sum()

# Calculate the percentage for each value in the column
Install_OR_NoInstall['Percentage'] = Install_OR_NoInstall['Raw User Id'].apply(lambda x: (x / column_total) * 100)

In [58]:
Install_OR_NoInstall

Raw User Id  Percentage
installs                         
0.0            120873   99.016982
1.0              1200    0.983018

*Above we can clearly see that the data when there was install is neglibile, not even 1%.<br>Whereas for no install it's 99%*<br>So, there is data imbalance

In [59]:
All_Raw_User_IDs_Summarized_ByAllValues=All_Merged_copy.groupby(['Raw User Id']).agg({'installs': 'sum','impressions': 'sum'})
All_Raw_User_IDs_Summarized_ByAllValues

installs  impressions
Raw User Id                                                
00000481-7D49-46DF-AF21-A8A2AAD60930       0.0          3.0
000020F8-F3AF-41F8-8896-82F5415E4ACB       0.0          1.0
0001826B-3F99-4612-81B3-DB7CDD60AEC8       0.0          2.0
0003FA5B-77BB-4F02-B625-0F16C990E304       0.0          1.0
00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       0.0          1.0
...                                        ...          ...
f4b4af4e-6eb6-4dd0-a6da-cd7c1caa3dbc       0.0          2.0
f69dade6-5469-45e2-9799-df163940c0ab       0.0          6.0
f7baef29-3123-46d3-87ad-45a9f14a8227       0.0          6.0
fa5c0287-8958-45d0-997a-cd591b483756       0.0         60.0
fc85a4d7-e7fc-4794-8f49-f617fad8821f       0.0          6.0

[122073 rows x 2 columns]

In [60]:
All_Raw_User_IDs_Summarized_ByAllValues.reset_index(inplace=True)
All_Raw_User_IDs_Summarized_ByAllValues

Raw User Id  installs  impressions
0       00000481-7D49-46DF-AF21-A8A2AAD60930       0.0          3.0
1       000020F8-F3AF-41F8-8896-82F5415E4ACB       0.0          1.0
2       0001826B-3F99-4612-81B3-DB7CDD60AEC8       0.0          2.0
3       0003FA5B-77BB-4F02-B625-0F16C990E304       0.0          1.0
4       00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       0.0          1.0
...                                      ...       ...          ...
122068  f4b4af4e-6eb6-4dd0-a6da-cd7c1caa3dbc       0.0          2.0
122069  f69dade6-5469-45e2-9799-df163940c0ab       0.0          6.0
122070  f7baef29-3123-46d3-87ad-45a9f14a8227       0.0          6.0
122071  fa5c0287-8958-45d0-997a-cd591b483756       0.0         60.0
122072  fc85a4d7-e7fc-4794-8f49-f617fad8821f       0.0          6.0

[122073 rows x 3 columns]

In [61]:
# Filter the data for installs = 1
NoInstalls_Raw_User_IDs_Summarized_ByAllValues = All_Raw_User_IDs_Summarized_ByAllValues[All_Raw_User_IDs_Summarized_ByAllValues['installs'] == 0]

# Display the filtered DataFrame
print(NoInstalls_Raw_User_IDs_Summarized_ByAllValues)

                                 Raw User Id  installs  impressions
0       00000481-7D49-46DF-AF21-A8A2AAD60930       0.0          3.0
1       000020F8-F3AF-41F8-8896-82F5415E4ACB       0.0          1.0
2       0001826B-3F99-4612-81B3-DB7CDD60AEC8       0.0          2.0
3       0003FA5B-77BB-4F02-B625-0F16C990E304       0.0          1.0
4       00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       0.0          1.0
...                                      ...       ...          ...
122068  f4b4af4e-6eb6-4dd0-a6da-cd7c1caa3dbc       0.0          2.0
122069  f69dade6-5469-45e2-9799-df163940c0ab       0.0          6.0
122070  f7baef29-3123-46d3-87ad-45a9f14a8227       0.0          6.0
122071  fa5c0287-8958-45d0-997a-cd591b483756       0.0         60.0
122072  fc85a4d7-e7fc-4794-8f49-f617fad8821f       0.0          6.0

[120873 rows x 3 columns]


In [62]:
# Calculate the frequency of each unique impression value
impression_counts = All_Raw_User_IDs_Summarized_ByAllValues['impressions'].value_counts()

# Create a new DataFrame to summarize the values
Install_Raw_User_IDs_Summarized_Impression_Frequency = pd.DataFrame({'Impressions': impression_counts.index, 'Frequency': impression_counts.values})

# Sort the DataFrame by the 'Impressions' column
Install_Raw_User_IDs_Summarized_Impression_Frequency = Install_Raw_User_IDs_Summarized_Impression_Frequency.sort_values('Impressions')

# Display the summary DataFrame
print(Install_Raw_User_IDs_Summarized_Impression_Frequency)

     Impressions  Frequency
17           0.0        103
0            1.0      72626
1            2.0      24231
2            3.0       9535
3            4.0       5087
..           ...        ...
70         159.0          3
113        164.0          1
106        166.0          1
88         177.0          1
87         290.0          1

[116 rows x 2 columns]


In [63]:
# Order the DataFrame by a specific column in descending order
Install_Raw_User_IDs_Summarized_Impression_Frequency = Install_Raw_User_IDs_Summarized_Impression_Frequency.sort_values('Frequency', ascending=False)

# Display the ordered DataFrame
print(Install_Raw_User_IDs_Summarized_Impression_Frequency)

     Impressions  Frequency
0            1.0      72626
1            2.0      24231
2            3.0       9535
3            4.0       5087
4            5.0       2938
..           ...        ...
109         94.0          1
95          93.0          1
94          90.0          1
93          76.0          1
87         290.0          1

[116 rows x 2 columns]


In [64]:
# Calculate the column total
column_total = Install_Raw_User_IDs_Summarized_Impression_Frequency['Frequency'].sum()

# Calculate the percentage for each value in the column
Install_Raw_User_IDs_Summarized_Impression_Frequency['Percentage'] = Install_Raw_User_IDs_Summarized_Impression_Frequency['Frequency'].apply(lambda x: (x / column_total) * 100)

In [65]:
Install_Raw_User_IDs_Summarized_Impression_Frequency

Impressions  Frequency  Percentage
0            1.0      72626   59.493909
1            2.0      24231   19.849598
2            3.0       9535    7.810900
3            4.0       5087    4.167179
4            5.0       2938    2.406757
..           ...        ...         ...
109         94.0          1    0.000819
95          93.0          1    0.000819
94          90.0          1    0.000819
93          76.0          1    0.000819
87         290.0          1    0.000819

[116 rows x 3 columns]

In [66]:
# Calculate the weighted average
weighted_avg = np.average(Install_Raw_User_IDs_Summarized_Impression_Frequency['Impressions'], weights=Install_Raw_User_IDs_Summarized_Impression_Frequency['Frequency'])

# Print the weighted average
print(weighted_avg)

2.277940248867481


In [67]:
# Exclude the first two indices from values and weights
values_excluded = Install_Raw_User_IDs_Summarized_Impression_Frequency['Impressions'][2:]
weights_excluded = Install_Raw_User_IDs_Summarized_Impression_Frequency['Frequency'][2:]

# Calculate the total weight
total_weight = np.sum(weights_excluded)

# Calculate the weighted average
weighted_average = np.average(values_excluded, weights=weights_excluded)

print("1st 80% Excluded Weighted Average:", weighted_average)

1st 80% Excluded Weighted Average: 6.225690038071066


<font size="3">As per above EDA if we show (6 impressions per Raw User Id per 5 Day)<br>Then the probability of no installation is negligible.<br>So optimal frequency cap(impressions we show per user per day) can be set to 1 or 2.</font>

In [68]:
For_Segmentation = stark_impressions.groupby(['Raw User Id','CAMPAIGN_ID','IAB_CATEGORY_ID1','EXCHANGE_ID','REQ_COUNTRY_ID']).agg({'impressions': 'sum'})
For_Segmentation

impressions
Raw User Id                          CAMPAIGN_ID IAB_CATEGORY_ID1 EXCHANGE_ID REQ_COUNTRY_ID             
000020F8-F3AF-41F8-8896-82F5415E4ACB 134047      157.0            23          246                       1
0003FA5B-77BB-4F02-B625-0F16C990E304 134795      157.0            23          246                       1
00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6 133833      222.0            23          246                       1
00084AB4-C963-4924-B8E6-FD0D69C7A1BF 133833      222.0            23          246                       1
000881C8-A3AB-4D48-811C-D1876B0D0BD9 133830      187.0            10049       246                       1
...                                                                                                   ...
fa5c0287-8958-45d0-997a-cd591b483756 135018      11.0             10047       246                      26
                                                 43.0             10047       246                       1
                                                 157.0            34          246                       8
                                                                  10047       246                      18
fc85a4d7-e7fc-4794-8f49-f617fad8821f 135018      43.0             10047       246                       6

[60898 rows x 1 columns]

In [69]:
stark_impressions.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 278075 entries, 0 to 279920
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   REQ_COUNTRY_ID    278075 non-null  int64  
 1   Raw User Id       278075 non-null  object 
 2   App Id            274763 non-null  object 
 3   EXCHANGE_ID       278075 non-null  int64  
 4   DEVICE_OS         278075 non-null  int64  
 5   CAMPAIGN_ID       278075 non-null  int64  
 6   CREATIVE_ID_1     278075 non-null  int64  
 7   Unique Id         278075 non-null  object 
 8   Epoch Timestamp   278075 non-null  int64  
 9   IAB_CATEGORY_ID1  107601 non-null  float64
 10  impressions       278075 non-null  int64  
dtypes: float64(1), int64(7), object(3)
memory usage: 25.5+ MB


In [70]:
For_Segmentation.reset_index(inplace=True)
For_Segmentation

Raw User Id  CAMPAIGN_ID  IAB_CATEGORY_ID1  \
0      000020F8-F3AF-41F8-8896-82F5415E4ACB       134047             157.0   
1      0003FA5B-77BB-4F02-B625-0F16C990E304       134795             157.0   
2      00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       133833             222.0   
3      00084AB4-C963-4924-B8E6-FD0D69C7A1BF       133833             222.0   
4      000881C8-A3AB-4D48-811C-D1876B0D0BD9       133830             187.0   
...                                     ...          ...               ...   
60893  fa5c0287-8958-45d0-997a-cd591b483756       135018              11.0   
60894  fa5c0287-8958-45d0-997a-cd591b483756       135018              43.0   
60895  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60896  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60897  fc85a4d7-e7fc-4794-8f49-f617fad8821f       135018              43.0   

       EXCHANGE_ID  REQ_COUNTRY_ID  impressions  
0               23             246            1  
1               23             246            1  
2               23             246            1  
3               23             246            1  
4            10049             246            1  
...            ...             ...          ...  
60893        10047             246           26  
60894        10047             246            1  
60895           34             246            8  
60896        10047             246           18  
60897        10047             246            6  

[60898 rows x 6 columns]

In [71]:
All_For_Segmentation=pd.merge(For_Segmentation,RawUserId_Clicks, on=['Raw User Id'], how='left')

In [72]:
All_For_Segmentation=pd.merge(All_For_Segmentation,RawUserId_installs, on=['Raw User Id'], how='left')

In [73]:
All_For_Segmentation

Raw User Id  CAMPAIGN_ID  IAB_CATEGORY_ID1  \
0      000020F8-F3AF-41F8-8896-82F5415E4ACB       134047             157.0   
1      0003FA5B-77BB-4F02-B625-0F16C990E304       134795             157.0   
2      00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       133833             222.0   
3      00084AB4-C963-4924-B8E6-FD0D69C7A1BF       133833             222.0   
4      000881C8-A3AB-4D48-811C-D1876B0D0BD9       133830             187.0   
...                                     ...          ...               ...   
60893  fa5c0287-8958-45d0-997a-cd591b483756       135018              11.0   
60894  fa5c0287-8958-45d0-997a-cd591b483756       135018              43.0   
60895  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60896  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60897  fc85a4d7-e7fc-4794-8f49-f617fad8821f       135018              43.0   

       EXCHANGE_ID  REQ_COUNTRY_ID  impressions  Clicks  installs  
0               23             246            1     NaN       NaN  
1               23             246            1     NaN       NaN  
2               23             246            1     NaN       NaN  
3               23             246            1     NaN       NaN  
4            10049             246            1     NaN       NaN  
...            ...             ...          ...     ...       ...  
60893        10047             246           26     NaN       NaN  
60894        10047             246            1     NaN       NaN  
60895           34             246            8     NaN       NaN  
60896        10047             246           18     NaN       NaN  
60897        10047             246            6     NaN       NaN  

[60898 rows x 8 columns]

In [74]:
All_For_Segmentation = All_For_Segmentation.fillna({'Clicks': 0, 'impressions': 0, 'installs': 0})

In [75]:
All_For_Segmentation

Raw User Id  CAMPAIGN_ID  IAB_CATEGORY_ID1  \
0      000020F8-F3AF-41F8-8896-82F5415E4ACB       134047             157.0   
1      0003FA5B-77BB-4F02-B625-0F16C990E304       134795             157.0   
2      00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       133833             222.0   
3      00084AB4-C963-4924-B8E6-FD0D69C7A1BF       133833             222.0   
4      000881C8-A3AB-4D48-811C-D1876B0D0BD9       133830             187.0   
...                                     ...          ...               ...   
60893  fa5c0287-8958-45d0-997a-cd591b483756       135018              11.0   
60894  fa5c0287-8958-45d0-997a-cd591b483756       135018              43.0   
60895  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60896  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60897  fc85a4d7-e7fc-4794-8f49-f617fad8821f       135018              43.0   

       EXCHANGE_ID  REQ_COUNTRY_ID  impressions  Clicks  installs  
0               23             246            1     0.0       0.0  
1               23             246            1     0.0       0.0  
2               23             246            1     0.0       0.0  
3               23             246            1     0.0       0.0  
4            10049             246            1     0.0       0.0  
...            ...             ...          ...     ...       ...  
60893        10047             246           26     0.0       0.0  
60894        10047             246            1     0.0       0.0  
60895           34             246            8     0.0       0.0  
60896        10047             246           18     0.0       0.0  
60897        10047             246            6     0.0       0.0  

[60898 rows x 8 columns]

In [76]:
country_mapping = pd.read_excel('country_mapping.xlsx')
country_mapping.head()

REQ_COUNTRY_ID    COUNTRY_NAME
0               1            Null
1               2         Unknown
2              11     Afghanistan
3              12  Ã…land Islands
4              13         Albania

In [77]:
All_For_Segmentation=pd.merge(All_For_Segmentation,country_mapping, on=['REQ_COUNTRY_ID'], how='left')

In [78]:
All_For_Segmentation

Raw User Id  CAMPAIGN_ID  IAB_CATEGORY_ID1  \
0      000020F8-F3AF-41F8-8896-82F5415E4ACB       134047             157.0   
1      0003FA5B-77BB-4F02-B625-0F16C990E304       134795             157.0   
2      00044080-AE6D-4C18-9AF2-2B3EB4F8ECB6       133833             222.0   
3      00084AB4-C963-4924-B8E6-FD0D69C7A1BF       133833             222.0   
4      000881C8-A3AB-4D48-811C-D1876B0D0BD9       133830             187.0   
...                                     ...          ...               ...   
60893  fa5c0287-8958-45d0-997a-cd591b483756       135018              11.0   
60894  fa5c0287-8958-45d0-997a-cd591b483756       135018              43.0   
60895  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60896  fa5c0287-8958-45d0-997a-cd591b483756       135018             157.0   
60897  fc85a4d7-e7fc-4794-8f49-f617fad8821f       135018              43.0   

       EXCHANGE_ID  REQ_COUNTRY_ID  impressions  Clicks  installs  \
0               23             246            1     0.0       0.0   
1               23             246            1     0.0       0.0   
2               23             246            1     0.0       0.0   
3               23             246            1     0.0       0.0   
4            10049             246            1     0.0       0.0   
...            ...             ...          ...     ...       ...   
60893        10047             246           26     0.0       0.0   
60894        10047             246            1     0.0       0.0   
60895           34             246            8     0.0       0.0   
60896        10047             246           18     0.0       0.0   
60897        10047             246            6     0.0       0.0   

        COUNTRY_NAME  
0      United States  
1      United States  
2      United States  
3      United States  
4      United States  
...              ...  
60893  United States  
60894  United States  
60895  United States  
60896  United States  
60897  United States  

[60898 rows x 9 columns]

In [79]:
CountryWise_Segmentation = All_For_Segmentation.groupby(['COUNTRY_NAME']).agg({'impressions': 'sum','Clicks': 'sum','installs': 'sum'})
CountryWise_Segmentation

impressions  Clicks  installs
COUNTRY_NAME                                       
Afghanistan                    42     2.0       0.0
Australia                       1     0.0       0.0
Bahrain                         2     0.0       0.0
Belize                          2     0.0       0.0
Brazil                          9     0.0       0.0
Bulgaria                        2     0.0       0.0
Canada                        535    12.0       5.0
China                           9     1.0       0.0
Colombia                        1     0.0       0.0
Czech Republic                  1     0.0       0.0
Denmark                         5     0.0       0.0
Djibouti                        1     0.0       0.0
Egypt                           3     0.0       0.0
France                          4     0.0       0.0
Gambia                          1     0.0       0.0
Georgia                         3     0.0       0.0
Germany                      1696    45.0       4.0
Greece                          7     0.0       0.0
Guam                            4     0.0       0.0
Honduras                        4     6.0       0.0
India                           1     0.0       0.0
Indonesia                       5     0.0       0.0
Iraq                           92     7.0       4.0
Japan                          16     1.0       0.0
Jordan                         73     4.0       3.0
Korea, Republic of              2     0.0       0.0
Kuwait                          8     0.0       0.0
Latvia                          1     0.0       0.0
Luxembourg                      1     0.0       0.0
Malaysia                        2     0.0       0.0
Marshall Islands                2     0.0       0.0
Mexico                          4     0.0       1.0
Netherlands                    13    11.0       0.0
New Zealand                     1     0.0       0.0
Pakistan                        1     0.0       0.0
Philippines                    26     0.0       0.0
Puerto Rico                     4     0.0       0.0
Qatar                           2     1.0       1.0
Russian Federation              6     0.0       0.0
Saudi Arabia                    3     0.0       0.0
Seychelles                     57     9.0       3.0
Singapore                       9     0.0       0.0
Spain                           2     0.0       0.0
Thailand                        1     0.0       0.0
Turkey                          9     3.0       0.0
Ukraine                         7     0.0       0.0
United Arab Emirates            1     0.0       0.0
United Kingdom                  5     0.0       0.0
United States              104806  4096.0     495.0
Virgin Islands, U.S.            1     0.0       0.0

In [80]:
CountryWise_Segmentation.reset_index(inplace=True)
CountryWise_Segmentation

COUNTRY_NAME  impressions  Clicks  installs
0            Afghanistan           42     2.0       0.0
1              Australia            1     0.0       0.0
2                Bahrain            2     0.0       0.0
3                 Belize            2     0.0       0.0
4                 Brazil            9     0.0       0.0
5               Bulgaria            2     0.0       0.0
6                 Canada          535    12.0       5.0
7                  China            9     1.0       0.0
8               Colombia            1     0.0       0.0
9         Czech Republic            1     0.0       0.0
10               Denmark            5     0.0       0.0
11              Djibouti            1     0.0       0.0
12                 Egypt            3     0.0       0.0
13                France            4     0.0       0.0
14                Gambia            1     0.0       0.0
15               Georgia            3     0.0       0.0
16               Germany         1696    45.0       4.0
17                Greece            7     0.0       0.0
18                  Guam            4     0.0       0.0
19              Honduras            4     6.0       0.0
20                 India            1     0.0       0.0
21             Indonesia            5     0.0       0.0
22                  Iraq           92     7.0       4.0
23                 Japan           16     1.0       0.0
24                Jordan           73     4.0       3.0
25    Korea, Republic of            2     0.0       0.0
26                Kuwait            8     0.0       0.0
27                Latvia            1     0.0       0.0
28            Luxembourg            1     0.0       0.0
29              Malaysia            2     0.0       0.0
30      Marshall Islands            2     0.0       0.0
31                Mexico            4     0.0       1.0
32           Netherlands           13    11.0       0.0
33           New Zealand            1     0.0       0.0
34              Pakistan            1     0.0       0.0
35           Philippines           26     0.0       0.0
36           Puerto Rico            4     0.0       0.0
37                 Qatar            2     1.0       1.0
38    Russian Federation            6     0.0       0.0
39          Saudi Arabia            3     0.0       0.0
40            Seychelles           57     9.0       3.0
41             Singapore            9     0.0       0.0
42                 Spain            2     0.0       0.0
43              Thailand            1     0.0       0.0
44                Turkey            9     3.0       0.0
45               Ukraine            7     0.0       0.0
46  United Arab Emirates            1     0.0       0.0
47        United Kingdom            5     0.0       0.0
48         United States       104806  4096.0     495.0
49  Virgin Islands, U.S.            1     0.0       0.0

In [81]:
# Assuming 'df' is the DataFrame of interest
CountryWise_Segmentation['Click_Through_Rate'] = CountryWise_Segmentation['Clicks'] / CountryWise_Segmentation['impressions']

# Display the DataFrame with the calculated metric
CountryWise_Segmentation

COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate
0            Afghanistan           42     2.0       0.0            0.047619
1              Australia            1     0.0       0.0            0.000000
2                Bahrain            2     0.0       0.0            0.000000
3                 Belize            2     0.0       0.0            0.000000
4                 Brazil            9     0.0       0.0            0.000000
5               Bulgaria            2     0.0       0.0            0.000000
6                 Canada          535    12.0       5.0            0.022430
7                  China            9     1.0       0.0            0.111111
8               Colombia            1     0.0       0.0            0.000000
9         Czech Republic            1     0.0       0.0            0.000000
10               Denmark            5     0.0       0.0            0.000000
11              Djibouti            1     0.0       0.0            0.000000
12                 Egypt            3     0.0       0.0            0.000000
13                France            4     0.0       0.0            0.000000
14                Gambia            1     0.0       0.0            0.000000
15               Georgia            3     0.0       0.0            0.000000
16               Germany         1696    45.0       4.0            0.026533
17                Greece            7     0.0       0.0            0.000000
18                  Guam            4     0.0       0.0            0.000000
19              Honduras            4     6.0       0.0            1.500000
20                 India            1     0.0       0.0            0.000000
21             Indonesia            5     0.0       0.0            0.000000
22                  Iraq           92     7.0       4.0            0.076087
23                 Japan           16     1.0       0.0            0.062500
24                Jordan           73     4.0       3.0            0.054795
25    Korea, Republic of            2     0.0       0.0            0.000000
26                Kuwait            8     0.0       0.0            0.000000
27                Latvia            1     0.0       0.0            0.000000
28            Luxembourg            1     0.0       0.0            0.000000
29              Malaysia            2     0.0       0.0            0.000000
30      Marshall Islands            2     0.0       0.0            0.000000
31                Mexico            4     0.0       1.0            0.000000
32           Netherlands           13    11.0       0.0            0.846154
33           New Zealand            1     0.0       0.0            0.000000
34              Pakistan            1     0.0       0.0            0.000000
35           Philippines           26     0.0       0.0            0.000000
36           Puerto Rico            4     0.0       0.0            0.000000
37                 Qatar            2     1.0       1.0            0.500000
38    Russian Federation            6     0.0       0.0            0.000000
39          Saudi Arabia            3     0.0       0.0            0.000000
40            Seychelles           57     9.0       3.0            0.157895
41             Singapore            9     0.0       0.0            0.000000
42                 Spain            2     0.0       0.0            0.000000
43              Thailand            1     0.0       0.0            0.000000
44                Turkey            9     3.0       0.0            0.333333
45               Ukraine            7     0.0       0.0            0.000000
46  United Arab Emirates            1     0.0       0.0            0.000000
47        United Kingdom            5     0.0       0.0            0.000000
48         United States       104806  4096.0     495.0            0.039082
49  Virgin Islands, U.S.            1     0.0       0.0            0.000000

In [82]:
CountryWise_Segmentation['Click_Through_Rate'] = CountryWise_Segmentation['Click_Through_Rate'].fillna(0)

# Display the updated DataFrame
print(CountryWise_Segmentation)

            COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate
0            Afghanistan           42     2.0       0.0            0.047619
1              Australia            1     0.0       0.0            0.000000
2                Bahrain            2     0.0       0.0            0.000000
3                 Belize            2     0.0       0.0            0.000000
4                 Brazil            9     0.0       0.0            0.000000
5               Bulgaria            2     0.0       0.0            0.000000
6                 Canada          535    12.0       5.0            0.022430
7                  China            9     1.0       0.0            0.111111
8               Colombia            1     0.0       0.0            0.000000
9         Czech Republic            1     0.0       0.0            0.000000
10               Denmark            5     0.0       0.0            0.000000
11              Djibouti            1     0.0       0.0            0.000000
12          

In [83]:
# Assuming 'df' is the DataFrame of interest
CountryWise_Segmentation['click to install rate'] = CountryWise_Segmentation['installs'] / CountryWise_Segmentation['Clicks']

# Display the DataFrame with the calculated metric
CountryWise_Segmentation

COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate  \
0            Afghanistan           42     2.0       0.0            0.047619   
1              Australia            1     0.0       0.0            0.000000   
2                Bahrain            2     0.0       0.0            0.000000   
3                 Belize            2     0.0       0.0            0.000000   
4                 Brazil            9     0.0       0.0            0.000000   
5               Bulgaria            2     0.0       0.0            0.000000   
6                 Canada          535    12.0       5.0            0.022430   
7                  China            9     1.0       0.0            0.111111   
8               Colombia            1     0.0       0.0            0.000000   
9         Czech Republic            1     0.0       0.0            0.000000   
10               Denmark            5     0.0       0.0            0.000000   
11              Djibouti            1     0.0       0.0            0.000000   
12                 Egypt            3     0.0       0.0            0.000000   
13                France            4     0.0       0.0            0.000000   
14                Gambia            1     0.0       0.0            0.000000   
15               Georgia            3     0.0       0.0            0.000000   
16               Germany         1696    45.0       4.0            0.026533   
17                Greece            7     0.0       0.0            0.000000   
18                  Guam            4     0.0       0.0            0.000000   
19              Honduras            4     6.0       0.0            1.500000   
20                 India            1     0.0       0.0            0.000000   
21             Indonesia            5     0.0       0.0            0.000000   
22                  Iraq           92     7.0       4.0            0.076087   
23                 Japan           16     1.0       0.0            0.062500   
24                Jordan           73     4.0       3.0            0.054795   
25    Korea, Republic of            2     0.0       0.0            0.000000   
26                Kuwait            8     0.0       0.0            0.000000   
27                Latvia            1     0.0       0.0            0.000000   
28            Luxembourg            1     0.0       0.0            0.000000   
29              Malaysia            2     0.0       0.0            0.000000   
30      Marshall Islands            2     0.0       0.0            0.000000   
31                Mexico            4     0.0       1.0            0.000000   
32           Netherlands           13    11.0       0.0            0.846154   
33           New Zealand            1     0.0       0.0            0.000000   
34              Pakistan            1     0.0       0.0            0.000000   
35           Philippines           26     0.0       0.0            0.000000   
36           Puerto Rico            4     0.0       0.0            0.000000   
37                 Qatar            2     1.0       1.0            0.500000   
38    Russian Federation            6     0.0       0.0            0.000000   
39          Saudi Arabia            3     0.0       0.0            0.000000   
40            Seychelles           57     9.0       3.0            0.157895   
41             Singapore            9     0.0       0.0            0.000000   
42                 Spain            2     0.0       0.0            0.000000   
43              Thailand            1     0.0       0.0            0.000000   
44                Turkey            9     3.0       0.0            0.333333   
45               Ukraine            7     0.0       0.0            0.000000   
46  United Arab Emirates            1     0.0       0.0            0.000000   
47        United Kingdom            5     0.0       0.0            0.000000   
48         United States       104806  4096.0     495.0            0.039082   
49  Virgin Islands, U.S.            1     0.0       0.0       

In [84]:
CountryWise_Segmentation['click to install rate'] = CountryWise_Segmentation['click to install rate'].fillna(0)

# Display the updated DataFrame
print(CountryWise_Segmentation)

            COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate  \
0            Afghanistan           42     2.0       0.0            0.047619   
1              Australia            1     0.0       0.0            0.000000   
2                Bahrain            2     0.0       0.0            0.000000   
3                 Belize            2     0.0       0.0            0.000000   
4                 Brazil            9     0.0       0.0            0.000000   
5               Bulgaria            2     0.0       0.0            0.000000   
6                 Canada          535    12.0       5.0            0.022430   
7                  China            9     1.0       0.0            0.111111   
8               Colombia            1     0.0       0.0            0.000000   
9         Czech Republic            1     0.0       0.0            0.000000   
10               Denmark            5     0.0       0.0            0.000000   
11              Djibouti            1     0.0       

In [85]:
# Assuming 'df' is the DataFrame of interest
CountryWise_Segmentation['Install_per_1000_Impression'] = CountryWise_Segmentation['Click_Through_Rate'] * CountryWise_Segmentation['click to install rate']

# Display the DataFrame with the calculated metric
CountryWise_Segmentation

COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate  \
0            Afghanistan           42     2.0       0.0            0.047619   
1              Australia            1     0.0       0.0            0.000000   
2                Bahrain            2     0.0       0.0            0.000000   
3                 Belize            2     0.0       0.0            0.000000   
4                 Brazil            9     0.0       0.0            0.000000   
5               Bulgaria            2     0.0       0.0            0.000000   
6                 Canada          535    12.0       5.0            0.022430   
7                  China            9     1.0       0.0            0.111111   
8               Colombia            1     0.0       0.0            0.000000   
9         Czech Republic            1     0.0       0.0            0.000000   
10               Denmark            5     0.0       0.0            0.000000   
11              Djibouti            1     0.0       0.0            0.000000   
12                 Egypt            3     0.0       0.0            0.000000   
13                France            4     0.0       0.0            0.000000   
14                Gambia            1     0.0       0.0            0.000000   
15               Georgia            3     0.0       0.0            0.000000   
16               Germany         1696    45.0       4.0            0.026533   
17                Greece            7     0.0       0.0            0.000000   
18                  Guam            4     0.0       0.0            0.000000   
19              Honduras            4     6.0       0.0            1.500000   
20                 India            1     0.0       0.0            0.000000   
21             Indonesia            5     0.0       0.0            0.000000   
22                  Iraq           92     7.0       4.0            0.076087   
23                 Japan           16     1.0       0.0            0.062500   
24                Jordan           73     4.0       3.0            0.054795   
25    Korea, Republic of            2     0.0       0.0            0.000000   
26                Kuwait            8     0.0       0.0            0.000000   
27                Latvia            1     0.0       0.0            0.000000   
28            Luxembourg            1     0.0       0.0            0.000000   
29              Malaysia            2     0.0       0.0            0.000000   
30      Marshall Islands            2     0.0       0.0            0.000000   
31                Mexico            4     0.0       1.0            0.000000   
32           Netherlands           13    11.0       0.0            0.846154   
33           New Zealand            1     0.0       0.0            0.000000   
34              Pakistan            1     0.0       0.0            0.000000   
35           Philippines           26     0.0       0.0            0.000000   
36           Puerto Rico            4     0.0       0.0            0.000000   
37                 Qatar            2     1.0       1.0            0.500000   
38    Russian Federation            6     0.0       0.0            0.000000   
39          Saudi Arabia            3     0.0       0.0            0.000000   
40            Seychelles           57     9.0       3.0            0.157895   
41             Singapore            9     0.0       0.0            0.000000   
42                 Spain            2     0.0       0.0            0.000000   
43              Thailand            1     0.0       0.0            0.000000   
44                Turkey            9     3.0       0.0            0.333333   
45               Ukraine            7     0.0       0.0            0.000000   
46  United Arab Emirates            1     0.0       0.0            0.000000   
47        United Kingdom            5     0.0       0.0            0.000000   
48         United States       104806  4096.0     495.0            0.039082   
49  Virgin Islands, U.S.            1     0.0       0.0       

In [86]:
CountryWise_Segmentation['Install_per_1000_Impression'] = CountryWise_Segmentation['Install_per_1000_Impression'].fillna(0)

# Display the updated DataFrame
print(CountryWise_Segmentation)

            COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate  \
0            Afghanistan           42     2.0       0.0            0.047619   
1              Australia            1     0.0       0.0            0.000000   
2                Bahrain            2     0.0       0.0            0.000000   
3                 Belize            2     0.0       0.0            0.000000   
4                 Brazil            9     0.0       0.0            0.000000   
5               Bulgaria            2     0.0       0.0            0.000000   
6                 Canada          535    12.0       5.0            0.022430   
7                  China            9     1.0       0.0            0.111111   
8               Colombia            1     0.0       0.0            0.000000   
9         Czech Republic            1     0.0       0.0            0.000000   
10               Denmark            5     0.0       0.0            0.000000   
11              Djibouti            1     0.0       

In [87]:
# Order the DataFrame by a specific column in descending order
CountryWise_Segmentation = CountryWise_Segmentation.sort_values('Install_per_1000_Impression', ascending=False)

# Display the ordered DataFrame
print(CountryWise_Segmentation)

            COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate  \
37                 Qatar            2     1.0       1.0            0.500000   
40            Seychelles           57     9.0       3.0            0.157895   
22                  Iraq           92     7.0       4.0            0.076087   
24                Jordan           73     4.0       3.0            0.054795   
6                 Canada          535    12.0       5.0            0.022430   
48         United States       104806  4096.0     495.0            0.039082   
16               Germany         1696    45.0       4.0            0.026533   
0            Afghanistan           42     2.0       0.0            0.047619   
29              Malaysia            2     0.0       0.0            0.000000   
30      Marshall Islands            2     0.0       0.0            0.000000   
31                Mexico            4     0.0       1.0            0.000000   
32           Netherlands           13    11.0       

In [88]:
CountryWise_Segmentation.head(15)

COUNTRY_NAME  impressions  Clicks  installs  Click_Through_Rate  \
37             Qatar            2     1.0       1.0            0.500000   
40        Seychelles           57     9.0       3.0            0.157895   
22              Iraq           92     7.0       4.0            0.076087   
24            Jordan           73     4.0       3.0            0.054795   
6             Canada          535    12.0       5.0            0.022430   
48     United States       104806  4096.0     495.0            0.039082   
16           Germany         1696    45.0       4.0            0.026533   
0        Afghanistan           42     2.0       0.0            0.047619   
29          Malaysia            2     0.0       0.0            0.000000   
30  Marshall Islands            2     0.0       0.0            0.000000   
31            Mexico            4     0.0       1.0            0.000000   
32       Netherlands           13    11.0       0.0            0.846154   
33       New Zealand            1     0.0       0.0            0.000000   
34          Pakistan            1     0.0       0.0            0.000000   
35       Philippines           26     0.0       0.0            0.000000   

    click to install rate  Install_per_1000_Impression  
37               1.000000                     0.500000  
40               0.333333                     0.052632  
22               0.571429                     0.043478  
24               0.750000                     0.041096  
6                0.416667                     0.009346  
48               0.120850                     0.004723  
16               0.088889                     0.002358  
0                0.000000                     0.000000  
29               0.000000                     0.000000  
30               0.000000                     0.000000  
31                    inf                     0.000000  
32               0.000000                     0.000000  
33               0.000000                     0.000000  
34               0.000000                     0.000000  
35               0.000000                     0.000000

As per the data except for Qatar, Seychelles, Iraq, Jordan, Canada, United States, Germany the probability of installation is 0%,<br>So I suggest focusing on these countries more and figuring out why of no installation in the other countries,<br>Also the 98% of the impressions are shown in US so we can make it 80% to 85% and delegate the remaining to the other countries where we are getting installation 

In [89]:
CampaignIDWise_Segmentation = All_For_Segmentation.groupby(['CAMPAIGN_ID']).agg({'Raw User Id': 'count'})
CampaignIDWise_Segmentation

Raw User Id
CAMPAIGN_ID             
132889              1044
133314                94
133387                20
133534               239
133535               142
133586              1115
133796               313
133830              5430
133833              8602
133853               273
134047             18760
134105              1783
134242               163
134255               564
134266               721
134686               758
134687                50
134795             14876
134851               340
134899               104
135018              5442
135093                 1
135096                33
135098                31

In [90]:
CampaignIDWise_Segmentation.reset_index(inplace=True)
CampaignIDWise_Segmentation.head()
#Raw User Id here is count of unique raw IDs which were shown those campaigns

CAMPAIGN_ID  Raw User Id
0       132889         1044
1       133314           94
2       133387           20
3       133534          239
4       133535          142

In [91]:
# Calculate the column total
column_total = CampaignIDWise_Segmentation['Raw User Id'].sum()

# Calculate the percentage for each value in the column
CampaignIDWise_Segmentation['Percentage'] = CampaignIDWise_Segmentation['Raw User Id'].apply(lambda x: (x / column_total) * 100)

In [92]:
CampaignIDWise_Segmentation

CAMPAIGN_ID  Raw User Id  Percentage
0        132889         1044    1.714342
1        133314           94    0.154356
2        133387           20    0.032842
3        133534          239    0.392460
4        133535          142    0.233177
5        133586         1115    1.830930
6        133796          313    0.513974
7        133830         5430    8.916549
8        133833         8602   14.125259
9        133853          273    0.448291
10       134047        18760   30.805609
11       134105         1783    2.927847
12       134242          163    0.267661
13       134255          564    0.926139
14       134266          721    1.183947
15       134686          758    1.244704
16       134687           50    0.082105
17       134795        14876   24.427732
18       134851          340    0.558311
19       134899          104    0.170777
20       135018         5442    8.936254
21       135093            1    0.001642
22       135096           33    0.054189
23       135098           31    0.050905

In [93]:
# Order the DataFrame by a specific column in descending order
CampaignIDWise_Segmentation = CampaignIDWise_Segmentation.sort_values('Percentage', ascending=False)

# Display the ordered DataFrame
print(CampaignIDWise_Segmentation)

    CAMPAIGN_ID  Raw User Id  Percentage
10       134047        18760   30.805609
17       134795        14876   24.427732
8        133833         8602   14.125259
20       135018         5442    8.936254
7        133830         5430    8.916549
11       134105         1783    2.927847
5        133586         1115    1.830930
0        132889         1044    1.714342
15       134686          758    1.244704
14       134266          721    1.183947
13       134255          564    0.926139
18       134851          340    0.558311
6        133796          313    0.513974
9        133853          273    0.448291
3        133534          239    0.392460
12       134242          163    0.267661
4        133535          142    0.233177
19       134899          104    0.170777
1        133314           94    0.154356
16       134687           50    0.082105
22       135096           33    0.054189
23       135098           31    0.050905
2        133387           20    0.032842
21       135093 

The 99% of the Raw User Ids had no installation in the dataset,<br>Out of which 54% were shown campaigns associated with CAMPAIGN_ID (134047, 134795) with (30%, 24%) weigtage respectively.<br>Without more data associated with installations, it's not safe to say that the no installations were solely because of these campaigns.<br>But we can atleast stop presenting 134047 campaign till further analysis.

In [94]:
CampaignIDWise_Segmentation = All_For_Segmentation.groupby(['CAMPAIGN_ID']).agg({'Raw User Id': 'count'})
CampaignIDWise_Segmentation

Raw User Id
CAMPAIGN_ID             
132889              1044
133314                94
133387                20
133534               239
133535               142
133586              1115
133796               313
133830              5430
133833              8602
133853               273
134047             18760
134105              1783
134242               163
134255               564
134266               721
134686               758
134687                50
134795             14876
134851               340
134899               104
135018              5442
135093                 1
135096                33
135098                31

In [95]:
EXCHANGE_IDWise_Segmentation = All_For_Segmentation.groupby(['EXCHANGE_ID']).agg({'Raw User Id': 'count'})
EXCHANGE_IDWise_Segmentation

Raw User Id
EXCHANGE_ID             
11                    96
12                   349
14                    13
18                   227
20                    18
23                 46952
34                    77
10031                786
10032                 63
10036                353
10038               1066
10045                 11
10046               1639
10047                219
10048                165
10049               8864

In [96]:
EXCHANGE_IDWise_Segmentation.reset_index(inplace=True)
EXCHANGE_IDWise_Segmentation.head()

EXCHANGE_ID  Raw User Id
0           11           96
1           12          349
2           14           13
3           18          227
4           20           18

In [97]:
# Calculate the column total
column_total = EXCHANGE_IDWise_Segmentation['Raw User Id'].sum()

# Calculate the percentage for each value in the column
EXCHANGE_IDWise_Segmentation['Percentage'] = EXCHANGE_IDWise_Segmentation['Raw User Id'].apply(lambda x: (x / column_total) * 100)

In [98]:
# Order the DataFrame by a specific column in descending order
EXCHANGE_IDWise_Segmentation = EXCHANGE_IDWise_Segmentation.sort_values('Percentage', ascending=False)

# Display the ordered DataFrame
print(EXCHANGE_IDWise_Segmentation)

    EXCHANGE_ID  Raw User Id  Percentage
5            23        46952   77.099412
15        10049         8864   14.555486
12        10046         1639    2.691386
10        10038         1066    1.750468
7         10031          786    1.290683
9         10036          353    0.579658
1            12          349    0.573089
3            18          227    0.372754
13        10047          219    0.359618
14        10048          165    0.270945
0            11           96    0.157641
6            34           77    0.126441
8         10032           63    0.103452
4            20           18    0.029558
2            14           13    0.021347
11        10045           11    0.018063


In [99]:
EXCHANGE_IDWise_Segmentation.head()

EXCHANGE_ID  Raw User Id  Percentage
5            23        46952   77.099412
15        10049         8864   14.555486
12        10046         1639    2.691386
10        10038         1066    1.750468
7         10031          786    1.290683

Ad Exchange associated with Exchange_ID 23 might be contributing to the no installation,<br>For that reason we should investigate further to figure out if that's actually what is happening and stop buying from there in that case. 

In [100]:
IAB_CATEGORY_Wise_Segmentation = All_For_Segmentation.groupby(['IAB_CATEGORY_ID1']).agg({'Raw User Id': 'count'})
IAB_CATEGORY_Wise_Segmentation

Raw User Id
IAB_CATEGORY_ID1             
11.0                    10686
12.0                      178
17.0                      402
19.0                       35
43.0                     1005
56.0                        1
68.0                      126
76.0                       19
84.0                      442
92.0                      219
123.0                       1
138.0                       6
157.0                   16170
162.0                     502
164.0                      14
180.0                      40
182.0                      53
187.0                    4862
205.0                     342
209.0                      42
222.0                   23563
223.0                       1
231.0                      56
241.0                     157
250.0                       9
295.0                      80
296.0                       2
302.0                     102
325.0                     395
331.0                     183
339.0                       8
371.0                       4
387.0                    1193

In [101]:
IAB_CATEGORY_Wise_Segmentation.reset_index(inplace=True)
IAB_CATEGORY_Wise_Segmentation.head()

IAB_CATEGORY_ID1  Raw User Id
0              11.0        10686
1              12.0          178
2              17.0          402
3              19.0           35
4              43.0         1005

In [102]:
# Calculate the column total
column_total = IAB_CATEGORY_Wise_Segmentation['Raw User Id'].sum()

# Calculate the percentage for each value in the column
IAB_CATEGORY_Wise_Segmentation['Percentage'] = IAB_CATEGORY_Wise_Segmentation['Raw User Id'].apply(lambda x: (x / column_total) * 100)

In [103]:
# Order the DataFrame by a specific column in descending order
IAB_CATEGORY_Wise_Segmentation = IAB_CATEGORY_Wise_Segmentation.sort_values('Percentage', ascending=False)

# Display the ordered DataFrame
print(IAB_CATEGORY_Wise_Segmentation)

    IAB_CATEGORY_ID1  Raw User Id  Percentage
20             222.0        23563   38.692568
12             157.0        16170   26.552596
0               11.0        10686   17.547374
17             187.0         4862    7.983842
32             387.0         1193    1.959013
4               43.0         1005    1.650301
13             162.0          502    0.824329
8               84.0          442    0.725804
2               17.0          402    0.660120
28             325.0          395    0.648626
18             205.0          342    0.561595
9               92.0          219    0.359618
29             331.0          183    0.300502
1               12.0          178    0.292292
23             241.0          157    0.257808
6               68.0          126    0.206903
27             302.0          102    0.167493
25             295.0           80    0.131367
22             231.0           56    0.091957
16             182.0           53    0.087031
19             209.0           42 